src/background/services/openPayments.ts

```diff
@@ -298,13 +326,60 @@ export class OpenPaymentsService {
       assetScale: walletAddress.assetScale
     })
 
+    await this.initClient(walletAddress.id)
+    await this.completeGrant(amount, walletAddress, recurring)
+
+    await this.storage.set({
+      walletAddress,
+      rateOfPay,
+      minRateOfPay,
+      maxRateOfPay,
+      connected: true
+    })
+  }
+
+  async addFunds({ amount, recurring }: AddFundsPayload) {
+    const { walletAddress } = await this.storage.get(['walletAddress'])
+
+    const grantDetails = await this.completeGrant(
+      amount,
+      walletAddress!,
+      recurring
+    )
+
+    // cancel existing grants of same type, if any
+    const grants = await this.storage.get(['oneTimeGrant', 'recurringGrant'])
+    if (grants.oneTimeGrant && !recurring) {
+      await this.cancelGrant(grants.oneTimeGrant.continue)
+    } else if (grants.recurringGrant && recurring) {
+      await this.cancelGrant(grants.recurringGrant.continue)
+    }
+
+    // If we already have a recurring grant, and user adds a one-time popup,
+    // prefer existing recurring grant. If the recurring grant doesn't have
+    // enough funds, which is likely the reason to add a one-time popup (but not
+    // always, as user can add top-up before running out of funds in future),
+    // we'll switch to the one-time grant when we try to make a payment.
```

Comment:
I understand your concern for the future, but right now it would make the most sense to automatically switch to `one-time` if the user already has a recurring grant and ran out of funds.

Reply:
Switch to whatever the new grant is, got it.